client/lib/scroll-to/index.js

```diff
@@ -17,34 +12,112 @@ function getCurrentScroll( container ) {
 		y = window.pageYOffset || document.documentElement.scrollTop;
-	return { x: x, y: y };
+	return { x, y };
 }
 
 function makeScrollUpdater( container ) {
 	container = container && container.scrollTop !== undefined ? container : window;
 
-	return function updateScroll() {
+	return function updateScroll( x, y ) {
 		if ( container === window ) {
-			container.scrollTo( this.x, this.y );
+			container.scrollTo( x, y );
 		} else {
-			container.scrollTop = this.y;
-			container.scrollLeft = this.x;
+			container.scrollTop = y;
+			container.scrollLeft = x;
 		}
 	};
 }
 
-function animate() {
-	if ( ! TWEEN.getAll().length ) {
-		return;
+class Stepper {
+	constructor( options ) {
+		this.x = options.x;
+		this.y = options.y;
+		this.startTime = null;
+		this.duration = options.duration;
+		this.start = options.start;
+		this.end = options.end;
+		this.onStart = options.onStart;
+		this.onComplete = options.onComplete;
+		this.updater = options.updater;
+		this.easing = options.easing;
+		this.container = options.container;
 	}
 
-	if ( 'undefined' !== typeof window && window.requestAnimationFrame ) {
-		window.requestAnimationFrame( animate );
-	} else {
-		defer( animate );
+	animate() {
+		this.nextFrame = requestAnimationFrame( this.step );
+	}
+
+	cancel() {
+		if ( this.nextFrame ) {
+			cancelAnimationFrame( this.nextFrame );
+			this.nextFrame = null;
+		}
+		if ( this.finishTimeout ) {
+			clearTimeout( this.finishTimeout );
+			this.finishTimeout = null;
+		}
+	}
+
+	jumpTo( x, y ) {
+		this.cancel();
+		this.end = {
+			x,
+			y,
+		};
+		this.updater( x, y );
+		this.finishTimeout = setTimeout( this.finish, this.duration );
+	}
+
+	finish = () => {
+		this.updater( this.end.x, this.end.y );
+		if ( this.onComplete ) {
+			this.onComplete();
+		}
+		scrollers.delete( this.container );
+	};
+
+	step = ts => {
+		this.nextFrame = null;
+		if ( ! this.startTime ) {
+			this.startTime = ts;
+			this.animate();
+			return;
+		}
+
+		// are we at or past our target duration?
+		// go to the last coordinate and finish up
+		if ( ts - this.startTime >= this.duration ) {
+			this.finish();
+			return;
+		}
+
+		// how far along are we as a percentage?
+		const progress = ( ts - this.startTime ) / this.duration;
+
+		// ease the progress
+		const easedProgress = this.easing( progress );
+
+		// figure out the new values
+		const newX = Math.round( this.start.x + ( this.end.x - this.start.x ) * easedProgress );
+		const newY = Math.round( this.start.y + ( this.end.y - this.start.y ) * easedProgress );
+
+		if ( newX !== this.x || newY !== this.y ) {
+			this.x = newX;
+			this.y = newY;
+			this.updater( newX, newY );
+		}
+
+		// set up the next turn
+		this.animate();
+	};
+}
+
+function circularInOutEasing( val ) {
+	if ( ( val *= 2 ) < 1 ) {
+		return -0.5 * ( Math.sqrt( 1 - val * val ) - 1 );
 	}
 
-	TWEEN.update();
+	return 0.5 * ( Math.sqrt( 1 - ( val -= 2 ) * val ) + 1 );
```

Comment:
Where does this formula come from and what shape it is supposed to produce? The way how `val` is mutated in mid-expression makes is harder to understand than neccessary.

Reply:
It's the circular inout easing from tween.js